### PR TITLE
[FLINK-24402][metric] Add a metric for back-pressure from the ChangelogStateBackend

### PR DESCRIPTION
**Pull request by fredia**
## What is the purpose of the change

Currently, the task back-pressured by the changelog state backend is shown as busy, this PR adds `ioBackPressuredTimePerSecond` metric and accumulates it into total `BackPressuredTimePerSecond`.

## Brief change log

- Add `ioBackPressuredTimePerSecond` metric
- Accumulates `ioBackPressuredTimePerSecond` into total `backPressuredTimePerSecond`

## Verifying this change

This change added tests and can be verified by `TaskIOMetricGroupTest#testTaskIOMetricGroup()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by fredia** [comment]
Hi @rkhachatryan, could you please help take a review?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:600]: The comment above is no longer true with this change, right?
When a task is waiting for changelog availability (but not blocked in writer), it will be reported as idle.

Won't such a change without adjusting the UI be confusing to the users?

cc: @pnowojski

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:600]: > When a task is waiting for changelog availability (but not blocked in writer), it will be reported as idle.

Are you sure? With the code in this PR as it is right now, I think task blocked on this changelog availability future, would be reported as back pressured.

> Won't such a change without adjusting the UI be confusing to the users?

As far as I can tell, this PR will move "waiting for changelog availability" from "busy" to "backpressured" in the UI. From this perspective, this should be fine in the UI (UI only displays `busy/backpressured/idle` and `backpressured = soft + hard`) % we are changing semantic pf those metric a bit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java:196]: Should this be added up to the backpressured time? IMO it fits better to the busyness.

If given subtask is spending tons of time waiting for the changelog, that's problem of that particular subtask, not a downstream subtask's problem. And that's more or less the distinction between busyness/backpressured states/times. In other words, as it is proposed currently, large waiting time for changelog would mess up with autoscalers. Subtask that's blocked waiting on changelog should be scaled up, so this time should be reported as busy time imo.

Moreover, as it is right now, aren't you sometimes accounting changelog time as backpressured (if this is "soft waiting for changelog" via availability future) and sometimes as busy ("hard waiting for changelog" when actually blocked in the changelog code).

File: docs/content/docs/ops/metrics.md
Comment (pnowojski) [inline docs/content/docs/ops/metrics.md:1723]: I think we should generalize this description a bit or change the name of the metric to something with `changelog` name.

Either instead of saying that this metric only reports changelog time, rename `ioBackPressuredTime` to `ioTime` (vide my other comment), describe it as "time spent by task doing IO operations, currently only takes into account changelog state backend."

or rename the metric to `changelogTime` or `changelogBusyTime`, and keep the very detailed description.

?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:600]: Yes, sorry, that's what I wanted to say: "backpressured", not "idle".

So should the documentation be updated then?
E.g. in https://nightlies.apache.org/flink/flink-docs-master/docs/ops/monitoring/back_pressure/

> Internally, back pressure is judged based on the availability of output buffers. If a task has no available output buffers, then that task is considered back pressured. Idleness, on the other hand, is determined by whether or not there is input available.

(and the highlighted comment)

Comment (fredia) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: Thanks for the review and analysis, I moved the `changelogBusyTime` into the busy time.

> So should the documentation be updated then?

Because we still treat `changelogBusyTime` as busy time, the documentation can stay as it is.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java]: Thanks for the detailed explanation and analysis, treating `changelogTime` as busy time is better to understand. I have moved the `changelogBusyTime`  into the busy time.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java:192]: You have to include this value at least in the `TaskIOMetricGroup#getBusyTimePerSecond` and `TaskIOMetricGroup#getAccumulatedBusyTime`. + add a test coverage for that. Otherwise `busy + idle + backpressured` won't add up to `1000` anymore and `changelogBusyTimeMsPerSecond` won't be included in the WebUI.

Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java:192]: Currently, `busy time = total time - idle time - backpressure time`, I think the `changelog busy time` will be automatically calculated into the `total busy time`, I added `testConsistencyOfTime()` to verify this.

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java:193]: Ahh, good point, you are right.

**Review by pnowojski — APPROVED** [review]
Thanks! LGTM assuming green build.

**Comment by fredia** [comment]
@flinkbot run azure

**Pull request merged**
